commit 54a85d462447c1cb8a1638578a7fd086350b4d2d
Author: Gerd Hoffmann <kraxel@redhat.com>
Date:   Tue Aug 26 14:16:30 2014 +0200

    vbe: make bochs dispi interface return the correct memory size with qxl
    
    VgaState->vram_size is the size of the pci bar.  In case of qxl not the
    whole pci bar can be used as vga framebuffer.  Add a new variable
    vbe_size to handle that case.  By default (if unset) it equals
    vram_size, but qxl can set vbe_size to something else.
    
    This makes sure VBE_DISPI_INDEX_VIDEO_MEMORY_64K returns correct results
    and sanity checks are done with the correct size too.
    
    Cc: qemu-stable@nongnu.org
    Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>
    Reviewed-by: Laszlo Ersek <lersek@redhat.com>

Index: qemu-1.4.0/hw/qxl.c
===================================================================
--- qemu-1.4.0.orig/hw/qxl.c	2013-02-16 00:05:35.000000000 +0100
+++ qemu-1.4.0/hw/qxl.c	2016-05-27 15:46:54.407713626 +0200
@@ -2059,6 +2059,7 @@
 
     qxl->id = 0;
     qxl_init_ramsize(qxl);
+    vga->vbe_size = qxl->vgamem_size;
     vga->vram_size_mb = qxl->vga.vram_size >> 20;
     vga_common_init(vga);
     vga_init(vga, pci_address_space(dev), pci_address_space_io(dev), false);
Index: qemu-1.4.0/hw/vga.c
===================================================================
--- qemu-1.4.0.orig/hw/vga.c	2016-05-27 15:45:48.898395853 +0200
+++ qemu-1.4.0/hw/vga.c	2016-05-27 15:46:54.484379499 +0200
@@ -620,7 +620,7 @@
             val = s->vbe_regs[s->vbe_index];
         }
     } else if (s->vbe_index == VBE_DISPI_INDEX_VIDEO_MEMORY_64K) {
-        val = s->vram_size / (64 * 1024);
+        val = s->vbe_size / (64 * 1024);
     } else {
         val = 0;
         if (!(s->vbe_regs[VBE_DISPI_INDEX_ENABLE] & VBE_DISPI_GETCAPS)) {
@@ -785,7 +785,7 @@
                     line_offset = w >> 1;
                 else
                     line_offset = w * ((s->vbe_regs[VBE_DISPI_INDEX_BPP] + 7) >> 3);
-                h = s->vram_size / line_offset;
+                h = s->vbe_size / line_offset;
                 /* XXX: support weird bochs semantics ? */
                 if (h < s->vbe_regs[VBE_DISPI_INDEX_YRES])
                     return;
@@ -2330,6 +2330,9 @@
         s->vram_size <<= 1;
     }
     s->vram_size_mb = s->vram_size >> 20;
+    if (!s->vbe_size) {
+        s->vbe_size = s->vram_size;
+    }
 
     s->is_vbe_vmstate = 1;
     memory_region_init_ram(&s->vram, "vga.vram", s->vram_size);
Index: qemu-1.4.0/hw/vga_int.h
===================================================================
--- qemu-1.4.0.orig/hw/vga_int.h	2016-05-10 16:03:35.255832122 +0200
+++ qemu-1.4.0/hw/vga_int.h	2016-05-27 15:46:54.551045476 +0200
@@ -107,6 +107,7 @@
     MemoryRegion vram_vbe;
     uint32_t vram_size;
     uint32_t vram_size_mb; /* property */
+    uint32_t vbe_size;
     uint32_t latch;
     MemoryRegion *chain4_alias;
     uint8_t sr_index;
